File: sha2wordlist.c
/*
 * Copyright (c) 2009 Kirei AB. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef __linux__
#include <bsd/string.h>
#endif

#include "sha2.h"

#define BUFLEN 16384

const char *wordlist_even[] = {
	"aardvark", "absurd", "accrue", "acme", "adrift",
	"adult", "afflict", "ahead", "aimless", "Algol",
	"allow", "alone", "ammo", "ancient", "apple",
	"artist", "assume", "Athens", "atlas", "Aztec",
	"baboon", "backfield", "backward", "banjo", "beaming",
	"bedlamp", "beehive", "beeswax", "befriend", "Belfast",
	"berserk", "billiard", "bison", "blackjack", "blockade",
	"blowtorch", "bluebird", "bombast", "bookshelf", "brackish",
	"breadline", "breakup", "brickyard", "briefcase", "Burbank",
	"button", "buzzard", "cement", "chairlift", "chatter",
	"checkup", "chisel", "choking", "chopper", "Christmas",
	"clamshell", "classic", "classroom", "cleanup", "clockwork",
	"cobra", "commence", "concert", "cowbell", "crackdown",
	"cranky", "crowfoot", "crucial", "crumpled", "crusade",
	"cubic", "dashboard", "deadbolt", "deckhand", "dogsled",
	"dragnet", "drainage", "dreadful", "drifter", "dropper",
	"drumbeat", "drunken", "Dupont", "dwelling", "eating",
	"edict", "egghead", "eightball", "endorse", "endow",
	"enlist", "erase", "escape", "exceed", "eyeglass",
	"eyetooth", "facial", "fallout", "flagpole", "flatfoot",
	"flytrap", "fracture", "framework", "freedom", "frighten",
	"gazelle", "Geiger", "glitter", "glucose", "goggles",
	"goldfish", "gremlin", "guidance", "hamlet", "highchair",
	"hockey", "indoors", "indulge", "inverse", "involve",
	"island", "jawbone", "keyboard", "kickoff", "kiwi",
	"klaxon", "locale", "lockup", "merit", "minnow",
	"miser", "Mohawk", "mural", "music", "necklace",
	"Neptune", "newborn", "nightbird", "Oakland", "obtuse",
	"offload", "optic", "orca", "payday", "peachy",
	"pheasant", "physique", "playhouse", "Pluto", "preclude",
	"prefer", "preshrunk", "printer", "prowler", "pupil",
	"puppy", "python", "quadrant", "quiver", "quota",
	"ragtime", "ratchet", "rebirth", "reform", "regain",
	"reindeer", "rematch", "repay", "retouch", "revenge",
	"reward", "rhythm", "ribcage", "ringbolt", "robust",
	"rocker", "ruffled", "sailboat", "sawdust", "scallion",
	"scenic", "scorecard", "Scotland", "seabird", "select",
	"sentence", "shadow", "shamrock", "showgirl", "skullcap",
	"skydive", "slingshot", "slowdown", "snapline", "snapshot",
	"snowcap", "snowslide", "solo", "southward", "soybean",
	"spaniel", "spearhead", "spellbind", "spheroid", "spigot",
	"spindle", "spyglass", "stagehand", "stagnate", "stairway",
	"standard", "stapler", "steamship", "sterling", "stockman",
	"stopwatch", "stormy", "sugar", "surmount", "suspense",
	"sweatband", "swelter", "tactics", "talon", "tapeworm",
	"tempest", "tiger", "tissue", "tonic", "topmost",
	"tracker", "transit", "trauma", "treadmill", "Trojan",
	"trouble", "tumor", "tunnel", "tycoon", "uncut",
	"unearth", "unwind", "uproot", "upset", "upshot",
	"vapor", "village", "virus", "Vulcan", "waffle",
	"wallet", "watchword", "wayside", "willow", "woodlark",
	"Zulu"
};

const char *wordlist_odd[] = {
	"adroitness", "adviser", "aftermath", "aggregate", "alkali",
	"almighty", "amulet", "amusement", "antenna", "applicant",
	"Apollo", "armistice", "article", "asteroid", "Atlantic",
	"atmosphere", "autopsy", "Babylon", "backwater", "barbecue",
	"belowground", "bifocals", "bodyguard", "bookseller",
	"borderline", "bottomless", "Bradbury", "bravado",
	"Brazilian", "breakaway", "Burlington", "businessman",
	"butterfat", "Camelot", "candidate", "cannonball",
	"Capricorn", "caravan", "caretaker", "celebrate",
	"cellulose", "certify", "chambermaid", "Cherokee",
	"Chicago", "clergyman", "coherence", "combustion",
	"commando", "company", "component", "concurrent",
	"confidence", "conformist", "congregate", "consensus",
	"consulting", "corporate", "corrosion", "councilman",
	"crossover", "crucifix", "cumbersome", "customer", "Dakota",
	"decadence", "December", "decimal", "designing", "detector",
	"detergent", "determine", "dictator", "dinosaur",
	"direction", "disable", "disbelief", "disruptive",
	"distortion", "document", "embezzle", "enchanting",
	"enrollment", "enterprise", "equation", "equipment",
	"escapade", "Eskimo", "everyday", "examine", "existence",
	"exodus", "fascinate", "filament", "finicky", "forever",
	"fortitude", "frequency", "gadgetry", "Galveston",
	"getaway", "glossary", "gossamer", "graduate", "gravity",
	"guitarist", "hamburger", "Hamilton", "handiwork",
	"hazardous", "headwaters", "hemisphere", "hesitate",
	"hideaway", "holiness", "hurricane", "hydraulic",
	"impartial", "impetus", "inception", "indigo", "inertia",
	"infancy", "inferno", "informant", "insincere", "insurgent",
	"integrate", "intention", "inventive", "Istanbul",
	"Jamaica", "Jupiter", "leprosy", "letterhead", "liberty",
	"maritime", "matchmaker", "maverick", "Medusa", "megaton",
	"microscope", "microwave", "midsummer", "millionaire",
	"miracle", "misnomer", "molasses", "molecule", "Montana",
	"monument", "mosquito", "narrative", "nebula", "newsletter",
	"Norwegian", "October", "Ohio", "onlooker", "opulent",
	"Orlando", "outfielder", "Pacific", "pandemic", "Pandora",
	"paperweight", "paragon", "paragraph", "paramount",
	"passenger", "pedigree", "Pegasus", "penetrate",
	"perceptive", "performance", "pharmacy", "phonetic",
	"photograph", "pioneer", "pocketful", "politeness",
	"positive", "potato", "processor", "provincial",
	"proximate", "puberty", "publisher", "pyramid", "quantity",
	"racketeer", "rebellion", "recipe", "recover", "repellent",
	"replica", "reproduce", "resistor", "responsive",
	"retraction", "retrieval", "retrospect", "revenue",
	"revival", "revolver", "sandalwood", "sardonic", "Saturday",
	"savagery", "scavenger", "sensation", "sociable",
	"souvenir", "specialist", "speculate", "stethoscope",
	"stupendous", "supportive", "surrender", "suspicious",
	"sympathy", "tambourine", "telephone", "therapist",
	"tobacco", "tolerance", "tomorrow", "torpedo", "tradition",
	"travesty", "trombonist", "truncated", "typewriter",
	"ultimate", "undaunted", "underfoot", "unicorn", "unify",
	"universe", "unravel", "upcoming", "vacancy", "vagabond",
	"vertigo", "Virginia", "visitor", "vocalist", "voyager",
	"warranty", "Waterloo", "whimsical", "Wichita",
	"Wilmington", "Wyoming", "yesteryear", "Yucatan"
};

int
hex2int(char hc)
{
	if (hc >= '0' && hc <= '9')
		return hc - '0';
	if (hc >= 'A' && hc <= 'F')
		return hc - 'A' + 10;
	if (hc >= 'a' && hc <= 'f')
		return hc - 'a' + 10;
	return 0;
}

void
pgp_wordlist(const char *str, char *buf, size_t bufsize)
{
	int i;
	int b;
	size_t nwords;

	nwords = strlen(str) / 2;
	buf[0] = '\0';
	
	for (i = 0; i < nwords; i++) {
		b = (hex2int(str[i*2]) << 4) + hex2int(str[i*2+1]);
		if (i % 2)
			strlcat(buf, wordlist_odd[b], bufsize);
		else
			strlcat(buf, wordlist_even[b], bufsize);
		if (i != nwords-1)
			strlcat(buf, " ", bufsize);
	}
}

int
main(int argc, char **argv)
{
	int fd;
	ssize_t nbytes;
	SHA256_CTX ctx256;
	char hash[BUFLEN];
	char words[BUFLEN];

	fd = fileno(stdin);

	SHA256_Init(&ctx256);

	while ((nbytes = read(fd, hash, BUFLEN)) > 0)
		SHA256_Update(&ctx256, (unsigned char *) hash, nbytes);

	SHA256_End(&ctx256, hash);
	
	pgp_wordlist(hash, words, sizeof(words));

	printf("SHA-256:    %s\n",  hash);
	printf("PGP Words:  %s\n", words);

	exit(0);
}

File: sha2wordlist.py
"""
Copyright (c) 2009 Kirei AB. All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
"""

import hashlib
import sys
from typing import List

BUFLEN = 16384

# https://en.wikipedia.org/wiki/PGP_word_list
# https://philzimmermann.com/docs/PGP_word_list.pdf
WORDS = [
    ("aardvark", "adroitness"),
    ("absurd", "adviser"),
    ("accrue", "aftermath"),
    ("acme", "aggregate"),
    ("adrift", "alkali"),
    ("adult", "almighty"),
    ("afflict", "amulet"),
    ("ahead", "amusement"),
    ("aimless", "antenna"),
    ("Algol", "applicant"),
    ("allow", "Apollo"),
    ("alone", "armistice"),
    ("ammo", "article"),
    ("ancient", "asteroid"),
    ("apple", "Atlantic"),
    ("artist", "atmosphere"),
    ("assume", "autopsy"),
    ("Athens", "Babylon"),
    ("atlas", "backwater"),
    ("Aztec", "barbecue"),
    ("baboon", "belowground"),
    ("backfield", "bifocals"),
    ("backward", "bodyguard"),
    ("banjo", "bookseller"),
    ("beaming", "borderline"),
    ("bedlamp", "bottomless"),
    ("beehive", "Bradbury"),
    ("beeswax", "bravado"),
    ("befriend", "Brazilian"),
    ("Belfast", "breakaway"),
    ("berserk", "Burlington"),
    ("billiard", "businessman"),
    ("bison", "butterfat"),
    ("blackjack", "Camelot"),
    ("blockade", "candidate"),
    ("blowtorch", "cannonball"),
    ("bluebird", "Capricorn"),
    ("bombast", "caravan"),
    ("bookshelf", "caretaker"),
    ("brackish", "celebrate"),
    ("breadline", "cellulose"),
    ("breakup", "certify"),
    ("brickyard", "chambermaid"),
    ("briefcase", "Cherokee"),
    ("Burbank", "Chicago"),
    ("button", "clergyman"),
    ("buzzard", "coherence"),
    ("cement", "combustion"),
    ("chairlift", "commando"),
    ("chatter", "company"),
    ("checkup", "component"),
    ("chisel", "concurrent"),
    ("choking", "confidence"),
    ("chopper", "conformist"),
    ("Christmas", "congregate"),
    ("clamshell", "consensus"),
    ("classic", "consulting"),
    ("classroom", "corporate"),
    ("cleanup", "corrosion"),
    ("clockwork", "councilman"),
    ("cobra", "crossover"),
    ("commence", "crucifix"),
    ("concert", "cumbersome"),
    ("cowbell", "customer"),
    ("crackdown", "Dakota"),
    ("cranky", "decadence"),
    ("crowfoot", "December"),
    ("crucial", "decimal"),
    ("crumpled", "designing"),
    ("crusade", "detector"),
    ("cubic", "detergent"),
    ("dashboard", "determine"),
    ("deadbolt", "dictator"),
    ("deckhand", "dinosaur"),
    ("dogsled", "direction"),
    ("dragnet", "disable"),
    ("drainage", "disbelief"),
    ("dreadful", "disruptive"),
    ("drifter", "distortion"),
    ("dropper", "document"),
    ("drumbeat", "embezzle"),
    ("drunken", "enchanting"),
    ("Dupont", "enrollment"),
    ("dwelling", "enterprise"),
    ("eating", "equation"),
    ("edict", "equipment"),
    ("egghead", "escapade"),
    ("eightball", "Eskimo"),
    ("endorse", "everyday"),
    ("endow", "examine"),
    ("enlist", "existence"),
    ("erase", "exodus"),
    ("escape", "fascinate"),
    ("exceed", "filament"),
    ("eyeglass", "finicky"),
    ("eyetooth", "forever"),
    ("facial", "fortitude"),
    ("fallout", "frequency"),
    ("flagpole", "gadgetry"),
    ("flatfoot", "Galveston"),
    ("flytrap", "getaway"),
    ("fracture", "glossary"),
    ("framework", "gossamer"),
    ("freedom", "graduate"),
    ("frighten", "gravity"),
    ("gazelle", "guitarist"),
    ("Geiger", "hamburger"),
    ("glitter", "Hamilton"),
    ("glucose", "handiwork"),
    ("goggles", "hazardous"),
    ("goldfish", "headwaters"),
    ("gremlin", "hemisphere"),
    ("guidance", "hesitate"),
    ("hamlet", "hideaway"),
    ("highchair", "holiness"),
    ("hockey", "hurricane"),
    ("indoors", "hydraulic"),
    ("indulge", "impartial"),
    ("inverse", "impetus"),
    ("involve", "inception"),
    ("island", "indigo"),
    ("jawbone", "inertia"),
    ("keyboard", "infancy"),
    ("kickoff", "inferno"),
    ("kiwi", "informant"),
    ("klaxon", "insincere"),
    ("locale", "insurgent"),
    ("lockup", "integrate"),
    ("merit", "intention"),
    ("minnow", "inventive"),
    ("miser", "Istanbul"),
    ("Mohawk", "Jamaica"),
    ("mural", "Jupiter"),
    ("music", "leprosy"),
    ("necklace", "letterhead"),
    ("Neptune", "liberty"),
    ("newborn", "maritime"),
    ("nightbird", "matchmaker"),
    ("Oakland", "maverick"),
    ("obtuse", "Medusa"),
    ("offload", "megaton"),
    ("optic", "microscope"),
    ("orca", "microwave"),
    ("payday", "midsummer"),
    ("peachy", "millionaire"),
    ("pheasant", "miracle"),
    ("physique", "misnomer"),
    ("playhouse", "molasses"),
    ("Pluto", "molecule"),
    ("preclude", "Montana"),
    ("prefer", "monument"),
    ("preshrunk", "mosquito"),
    ("printer", "narrative"),
    ("prowler", "nebula"),
    ("pupil", "newsletter"),
    ("puppy", "Norwegian"),
    ("python", "October"),
    ("quadrant", "Ohio"),
    ("quiver", "onlooker"),
    ("quota", "opulent"),
    ("ragtime", "Orlando"),
    ("ratchet", "outfielder"),
    ("rebirth", "Pacific"),
    ("reform", "pandemic"),
    ("regain", "Pandora"),
    ("reindeer", "paperweight"),
    ("rematch", "paragon"),
    ("repay", "paragraph"),
    ("retouch", "paramount"),
    ("revenge", "passenger"),
    ("reward", "pedigree"),
    ("rhythm", "Pegasus"),
    ("ribcage", "penetrate"),
    ("ringbolt", "perceptive"),
    ("robust", "performance"),
    ("rocker", "pharmacy"),
    ("ruffled", "phonetic"),
    ("sailboat", "photograph"),
    ("sawdust", "pioneer"),
    ("scallion", "pocketful"),
    ("scenic", "politeness"),
    ("scorecard", "positive"),
    ("Scotland", "potato"),
    ("seabird", "processor"),
    ("select", "provincial"),
    ("sentence", "proximate"),
    ("shadow", "puberty"),
    ("shamrock", "publisher"),
    ("showgirl", "pyramid"),
    ("skullcap", "quantity"),
    ("skydive", "racketeer"),
    ("slingshot", "rebellion"),
    ("slowdown", "recipe"),
    ("snapline", "recover"),
    ("snapshot", "repellent"),
    ("snowcap", "replica"),
    ("snowslide", "reproduce"),
    ("solo", "resistor"),
    ("southward", "responsive"),
    ("soybean", "retraction"),
    ("spaniel", "retrieval"),
    ("spearhead", "retrospect"),
    ("spellbind", "revenue"),
    ("spheroid", "revival"),
    ("spigot", "revolver"),
    ("spindle", "sandalwood"),
    ("spyglass", "sardonic"),
    ("stagehand", "Saturday"),
    ("stagnate", "savagery"),
    ("stairway", "scavenger"),
    ("standard", "sensation"),
    ("stapler", "sociable"),
    ("steamship", "souvenir"),
    ("sterling", "specialist"),
    ("stockman", "speculate"),
    ("stopwatch", "stethoscope"),
    ("stormy", "stupendous"),
    ("sugar", "supportive"),
    ("surmount", "surrender"),
    ("suspense", "suspicious"),
    ("sweatband", "sympathy"),
    ("swelter", "tambourine"),
    ("tactics", "telephone"),
    ("talon", "therapist"),
    ("tapeworm", "tobacco"),
    ("tempest", "tolerance"),
    ("tiger", "tomorrow"),
    ("tissue", "torpedo"),
    ("tonic", "tradition"),
    ("topmost", "travesty"),
    ("tracker", "trombonist"),
    ("transit", "truncated"),
    ("trauma", "typewriter"),
    ("treadmill", "ultimate"),
    ("Trojan", "undaunted"),
    ("trouble", "underfoot"),
    ("tumor", "unicorn"),
    ("tunnel", "unify"),
    ("tycoon", "universe"),
    ("uncut", "unravel"),
    ("unearth", "upcoming"),
    ("unwind", "vacancy"),
    ("uproot", "vagabond"),
    ("upset", "vertigo"),
    ("upshot", "Virginia"),
    ("vapor", "visitor"),
    ("village", "vocalist"),
    ("virus", "voyager"),
    ("Vulcan", "warranty"),
    ("waffle", "Waterloo"),
    ("wallet", "whimsical"),
    ("watchword", "Wichita"),
    ("wayside", "Wilmington"),
    ("willow", "Wyoming"),
    ("woodlark", "yesteryear"),
    ("Zulu", "Yucatan"),
]


def pgp_wordlist(data: bytes) -> List[str]:
    """Translate bytes to list of PGP words."""
    odd = False
    words = []
    for byte in data:
        if odd:
            words.append(WORDS[byte][1])
        else:
            words.append(WORDS[byte][0])
        odd = not odd
    return words


def main() -> None:
    """Main function"""
    h = hashlib.new("sha256")

    while data:
        data = sys.stdin.buffer.read(BUFLEN)
        if data:
            h.update(data)
        else:
            break

    words = " ".join(pgp_wordlist(h.digest()))
    print("SHA-256:    ", h.hexdigest())
    print("PGP Words:  ", words)


if __name__ == "__main__":
    main()
